### ISSUE TITLE
Extending reductors

### ISSUE THREAD

**Issue by Reneg973** [ddemidov/vexcl#192]
For creating some statistics on integer (1 or 2 Bytes), I tried to use reductors but directly ran into the issue that they were not made for integers, but just for 'real', as the code tells.

So one of my first modifications was to extend reductor with an output type template parameter and renamed the 'real' type:
`template <typename InType, class RDC = SUM, typename OutType = InType>`
After modifying all the places where the OutType was required, I can now add a lot of shorts into a client defined larger type.

Now my code Looks liks:
`vex::reductor<short, vex::SUM, long long> sum;`
`auto sum = sum(a);`
`auto sum_sqr = sum(a*a);`

I then made a new reductor using 2 RDCs, like:
`template <typename InType, typename OutType, class RDC1, class RDC2>`, created a local combiner, which combines the code for both operations. Therefore I've splitted the getter into an initialization and the grid_stride_loop part.
The result now is:
`vex::reductor<short, double, vex::MIN, vex::MAX> min_max(ctx);`
`std::array<double,2> result = min_max(a);`
or
`vex::reductor<short, double, vex::SUM, vex::SUM_SQUARES> sum_sum2(ctx);`
`std::array<double,2> result = sum_sum2(a);`

But IMO all this is not generic enough, not satisfying :)
Very interesting would be a solution using variadic templates:
`template <typename InType, typename OutType, class...OPs>`. The Output could be a `std::tuple<>` or at least a `std::array<OutType, sizeof...(OPs)>`.

Would such a thing be possible somehow? And the other question, would it be practicable? I'm not very experienced with OpenCL kernels and don't know how much performance gain I could get when combining some reductor kernels into only one.

**Comment by ddemidov** [comment]
I am sorry for the inappropriately chosen template parameter name; the `real` type is in fact the output type for the reductor (the reductor takes arbitrary expressions as long as the result may be converted to the output type). So in the following example `sum` would reduce vector of `short`s and output an `int`:

```cpp
vex::Reductor<int> sum(ctx);
vex::vector<short> x(ctx, n);

int s = sum(x);
```

**Comment by ddemidov** [comment]
Regarding combination of operations, the case that you are interested in (gathering both sum of elements and sum of squares) may be achieved with the following helper function:
```cpp
VEX_FUNCTION(cl_long2, moments, (short, x),
  return (long2)(x, x*x);
);
vex::Reductor<cl_long2> sum(ctx);
vex::vector<short> x(ctx, n);
cl_long2 m12 = sum( moments(x) ); // Now m12.s[0] is sum(x), and m12.s[1] is sum(x*x)
```

**Comment by ddemidov** [comment]
The case of combining different reduction operations (min and max in your example) is more difficult. Currently, without changing the definition of `Reductor` class one could just introduce custom operation similar to [one of the builtins](https://github.com/ddemidov/vexcl/blob/master/vexcl/reductor.hpp#L108-L126).

But your idea of providing a generic solution that could make several reductions at once also looks interesting. I think it would be better to create a reduction operation wrapper that would allow to combine several operations for the same input type. Something like

```cpp
vex::Reductor<cl_double2, vex::Combine<vex::MIN, vex::MAX> > min_max(ctx);
```
I am not sure about the return type though (`cl_double2` here). The `Combine` class (or whatever it would be called) could probably define a C-analogue to `std::array<>` both on host and compute device sides of things.

**Comment by Reneg973** [comment]
I've compared the moments VEX_FUNCTION, which looked very promising, with my implementation and it's about 10% slower. I don't know the cause, also I don't have any high end OpenCL card. Nevertheless it's all about performance. I need a processing pipeline which is able to process at least 300 mega samples per second. On 4xCPU with a lot of SSE code we're able to reach that, with the disadvantage that the CPU is on very high load.
Also compared to a single core CPU implementation I just get a 5x Performance gain on a GTX 750Ti Card.

My code is:
`typedef float ConvType;`
`vex::vector<InType> a(ctx, A, CL_MEM_READ_ONLY);
`vex::vector<InType> b(ctx, B, CL_MEM_READ_ONLY);
`vex::Reductor<cl_ulong2, vex::SUM> sum(ctx);
`vex::Reductor<ConvType, vex::SUM_Kahan> sumK(ctx);
`VEX_FUNCTION(cl_ulong2, moments, (unsigned short, x), return (ulong2)(x, x*x););`
          auto normSize = 1.f / (w*h);
          prof.tic_cpu("start");
          using vex::tag;
          for (int i = 0; i < NUM_EXEC; ++i)
          {
            auto ta = tag<1>(a);
            auto mA = sum(moments(ta));
            auto meanA = mA.s[0];
            auto varianceA = normSize * mA.s[1] - meanA*meanA;
            auto tb = tag<2>(b);
            auto mB = sum(moments(tb));
            auto meanB = mB.s[0];
            auto varianceB = normSize * mB.s[1] - meanB*meanB;

            auto dStdDev12 = (normSize / (sqrt(varianceA*varianceB)));

            dCorr = sumK((ta - meanA) * (tb - meanB) * dStdDev12);
          }
          prof.tic_cpu("end");`

I have the feeling the GPU should be much faster, at least because I think that the data vector is just copied once. Am I wrong here?

**Comment by ddemidov** [comment]
> with my implementation and it's about 10% slower.

Did you implement it on top of VexCL? Can you share the code? You could try to compare the generated code for the two versions (by exporting VEXCL_SHOW_KERNELS=1 and running the program) and see if this helps to pinpoint the difference in performance.

What CPU did you compare performance with? GTX 750Ti has [86 GB/sec](http://www.geforce.com/hardware/desktop-gpus/geforce-gtx-750-ti/specifications) memory bandwidth, and the measured peak is at [68 GB/sec](https://github.com/krrishnarraj/clpeak/blob/master/results/NVIDIA_CUDA/GeForce_GTX_750.log#L9). An Intel i5 CPU peaks at [16 GB/sec](https://github.com/krrishnarraj/clpeak/blob/master/results/Intel(R)_OpenCL/Intel_Core_i5-4570.log#L9) which seems to be more or less consistent with your results.

**Comment by ddemidov** [comment]
#193 introduces `vex::CombineReductors<RDC1, RDC2, ...>` class that allows to combine several reductors into single kernel.

**Comment by Reneg973** [comment]
Thank you for the fast modifications.
Found a copy&paste error in the `CombineReductors<>::device_out::partial_reduce<>()`. It should be `Head::template impl<T>::device_out::name()`

**Comment by ddemidov** [comment]
Thanks for taking a moment to review the changes! Fixed in 05a28ed.

**Comment by Reneg973** [comment]
well I'm satisfied :)
